### ISSUE TITLE
RFE: TestRecorder UI must allow editing the generated test case

### ISSUE THREAD

**Issue by ThomasH99** [codenameone/CodenameOne#3039]
When creating a test case using the TestRecorder UI, some of the [Labels] and [TextAreas] generate checks on ALL labels or textareas, many of which are unlikely to be relevant for the test case at hand. E.g. in a screen you edit one of the displayed fields and you want to check that the edited value is correct, all other fields should be ignored

Currently the workflow to achieve this is to generate the test case, then edit the generated java to manually delete all the irrelevant tests, which require you to remember all the steps and all the screens of the test case. Making it near impossible in practice. 

To get around this, the generated test case code must be editable directly in the TestRecorder UI so you can make such edits on the fly.

**Comment by ThomasH99** [comment]
Seems to me that test TestRecorder is not really mature, which gives the impression it is not being used and that you may then not be interested in improving it. Please let me know your position, I don't want to waste time reporting errors and suggesting improvements if it's not going to be used and provide value for you or the community.

**Comment by codenameone** [comment]
This is recursive. If people don't use something we don't maintain it. Hence it's buggy and people don't use it. By its nature people use the recorder a few times and have the tests ready. Then they don't need it anymore so it deteriorates if people no longer need it.

**Comment by ThomasH99** [comment]
I agree. Checken and egg situation :-) I’m just asking because I’ve earlier made usability-related RFEs for features (I think on Network Monitor) and you didn’t seem to prioritize that kind of improvements. I’ll leave with this one for now and make more if it moves ahead.

PS. I don’t agree with your point on how people use a tool like TestRecorder. If it works, people continue to add tests as new features are added to an app and you regression-test the existing features (or improve coverage for existing features). I think TestRecorder is a great idea that just needs a bit more investment to reach the critical level of features/usabability. My guess is you’re not using it (extensively) yourself either.

**Comment by codenameone** [comment]
I suck at testing my work so personally I don't use it.
Steve mostly writes with his own framework mostly for testing the framework which is a pretty different task from testing an app. So neither one of us uses it. 

We're pretty busy with enterprise tasks so it's very hard to look past that. The main problem with the specific network monitor task is the fact that it's written in Swing so those changes require writing a lot of Swing code. Ideally we want to rewrite it completely using Codename One. But in order to do that we need to provide a way to run Codename One outside of the simulator without it colliding with the simulator itself yet with a way for it to impact the execution of the simulator. That's hard. So the main delay is the fact that we don't want to invest in something that we want to throw away.

**Comment by ThomasH99** [comment]
Thanks for the insight. I fully understand your focus on Enterprise customers, that benefits even those of us not yet reached that point yet by keeping CN1 moving forward. I’m less convinced by your intention to rewrite the UI in CN1. The other utilities where you’ve done that are not giving a very pleasant user experience and don’t really promote the value of CN1 very convincingly. They kind of, sort of, work, but the feeling is just too far behind native apps (based on my experience on Mac and Windows). And everyone will assume that’s the best that can be achieved with CN1 because coming from the CN1 team.

**Comment by codenameone** [comment]
IMO The main problem here is the lack of design for those features. I think that once our designer gets there we can significantly improve the UX/UI of these apps. He's a bit overworked with tasks at the moment. Swing is pretty much a dead end when it comes to UX or design by now.

**Comment by ThomasH99** [comment]
Design definitely counts, and I don’t have experience to recommend one framework or another.   But it has not been not just been about looks, bit also responsiveness and just ‘feeling native’ (there were several small things, can’t remember exactly what now). Maybe do a survey to hear from your user base?

**Comment by ThomasH99** [comment]
Not sure if the TestRecorder is something you still consider relevant, the idea is nice though so bumping this up. 

Good support for testing the UI of an app is definitely still useful and I haven't found any other good solutions for automating such tests (let me know if something exists that I've missed).

**Comment by liannacasper** [comment]
I tried working on this. The text is editable, I'm not sure how I can improve this behavior.  If you just want a small segment to be recorded just press record and stop recording when you don't want stuff to be recorded.
